### PR TITLE
s3proxy: add initial implementation

### PR DESCRIPTION
**Pull request by derpsteb**
### Context

**EDIT:** This PR now includes keyservice integration and proper envelope encryption (#2398).

This PR adds a new service to Constellation called s3proxy. It can be used to transparently encrypt and decrypt data when using AWS S3. Data that is created while using s3proxy is tagged with `constellation-encryption` and the body is encrypted using AES-256-GCM. Data that is retrieved while using s3proxy is decrypted if has an encrypted data encryption key in it's metadata. 

This PR is missing:
- e2e tests based on minio/mint: I would like to discuss again how we want to put this into the Constellation repo.
- deployment automation as a bazel target: plan is to have a target that builds the image, pushes it and deploys s3proxy with the image into a cluster that was setup separately. This seems like the easiest development setup. 
- ~integration with Constellation keyservice to derive keys: immediate next step.~

I decided to leave these things out to get some eyes on the implementation as it is now working with the e2e tests and Filestash.

There are no unittests so far as I didn't know what the correct behavior was until now. The implementation is validated through the minio/mint e2e tests and Filestash so far. Ideally I would like to revisit unittests after we have the MVP in the documentation and rely on e2e tests until then. 

### Proposed change(s)
- Add initial implementation of s3proxy

### Additional information
- ~Tagging as no-changelog as at least the keyservice integration is required before users should be pointed to this. That separate PR can link to this one to have the full story.~ Tagging as feature as this now includes the full MVP.
- Added Thomas as reviewer, but feedback from anyone else is welcome.

### Checklist

- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

File: s3proxy/internal/crypto/crypto.go
Comment (3u13r) [inline s3proxy/internal/crypto/crypto.go:8]: While this package's job seems simple, I'd suggest to use Tink's AEAD encryption: https://pkg.go.dev/github.com/google/tink/go/aead#example-package. This makes it even harder to make crypto mistakes.

File: s3proxy/internal/crypto/crypto.go
Comment (3u13r) [inline s3proxy/internal/crypto/crypto.go:67]: I think we should check the length of the ciphertext before those operations.

File: s3proxy/internal/router/object.go
Comment (3u13r) [inline s3proxy/internal/router/object.go:27]: I think we should address this right now. I'm very uneasy with variables that start with `testing` who not reside in a `_test.go` file. Especially if they are crypto keys. Especially if they are used right now.
Maybe since we aren't in a hurry with this feature and this is quite separate, let's wait for approval here and then create a PR with the KMS implementation to this branch?

File: s3proxy/internal/router/router.go
Comment (3u13r) [inline s3proxy/internal/router/router.go:69]: I'd refactor this switch case:
1. factor out the function calls like `containsBucket(req.Host)` to before the switch/case and instead assign those to bool variables which have a clear name. This should improve readability.
2. refactor each case "body" to its own function like `handle...()`.

File: s3proxy/internal/crypto/crypto.go
Comment (3u13r) [reply s3proxy/internal/crypto/crypto.go:8]: If we decide to still implement this ourselves, then we should at least add tests for this.

File: s3proxy/deploy/create_cert.sh
Comment (3u13r) [inline s3proxy/deploy/create_cert.sh:2]:
```suggestion
# $service is the name of the s3proxy service in kubernetes.
```
I think this makes it faster to parse out the variable. I don't know of any style guides for bash commenting tough.

**Review by 3u13r — CHANGES_REQUESTED** [review]

File: s3proxy/deploy/create_cert.sh
Comment (daniel-weisse) [inline s3proxy/deploy/create_cert.sh:1]: Why not use cert-manager to generate this certificate?

File: s3proxy/internal/crypto/crypto.go
Comment (derpsteb) [reply s3proxy/internal/crypto/crypto.go:8]: Addressed here: https://github.com/edgelesssys/constellation/pull/2398

File: s3proxy/internal/router/object.go
Comment (derpsteb) [reply s3proxy/internal/router/object.go:27]: Addressed here: https://github.com/edgelesssys/constellation/pull/2398

File: s3proxy/internal/crypto/crypto.go
Comment (derpsteb) [reply s3proxy/internal/crypto/crypto.go:29]: Removed in: https://github.com/edgelesssys/constellation/pull/2398

File: s3proxy/internal/crypto/crypto.go
Comment (derpsteb) [reply s3proxy/internal/crypto/crypto.go:67]: Removed in: https://github.com/edgelesssys/constellation/pull/2398

File: s3proxy/deploy/create_cert.sh
Comment (derpsteb) [reply s3proxy/deploy/create_cert.sh:2]: Replaced by cert-manager solution.

File: s3proxy/internal/crypto/crypto.go
Comment (derpsteb) [reply s3proxy/internal/crypto/crypto.go:36]: Removed in: https://github.com/edgelesssys/constellation/pull/2398

File: s3proxy/internal/router/router.go
Comment (derpsteb) [reply s3proxy/internal/router/router.go:69]: Code should a lot more concise now. Sorry for the convoluted initial version :see_no_evil:

File: s3proxy/cmd/main.go
Comment (derpsteb) [reply s3proxy/cmd/main.go:70]: Replaced port flag with `--no-tls` flag.

File: s3proxy/internal/router/router.go
Comment (daniel-weisse) [inline s3proxy/internal/router/router.go:359]:
```suggestion
func match(path, pattern string, vars ...any) bool {
```
nit

File: s3proxy/internal/router/router.go
Comment (daniel-weisse) [inline s3proxy/internal/router/router.go:376]: This function should not panic.
Return errors instead

File: s3proxy/internal/router/router.go
Comment (daniel-weisse) [inline s3proxy/internal/router/router.go:392]: This function is only used for 2 "hardcoded" patterns.
Additionally, this function may panic at runtime if someone decides to use this function with new patterns.
Instead of using this functions, compile the expressions as package scoped vars, and pass them to the `match` function.

E.g.:
```Go
var (
    // adjust naming
    matchKey = regexp.MustCompile("/(.+)")
    matchBucketAndKey = regexp.MustCompile("/([^/?]+)/(.+)")
)

func (r Router) Serve(w http.ResponseWriter, req *http.Request) {
    // ...
    if containsBucket(req.Host) {
        // BUCKET.s3.REGION.amazonaws.com
       parts := strings.Split(req.Host, ".")
       bucket = parts[0]

       matchingPath = match(req.URL.Path, matchKey, &key)
   } else {
       matchingPath = match(req.URL.Path, matchBucketAndKey, &bucket, &key)
    }
    // ...
}

func match((path, rxp *regexp.Regexp, vars ...any) bool {
    matches := rxp.FindStringSubmatch(path)
    // ...
}
```

**Review by daniel-weisse — CHANGES_REQUESTED** [review]

File: s3proxy/deploy/deployment-s3proxy.yaml
Comment (derpsteb) [reply s3proxy/deploy/deployment-s3proxy.yaml]: Not sure how to get an image from the constellation registry without merging.

File: s3proxy/internal/router/router.go
Comment (derpsteb) [reply s3proxy/internal/router/router.go]: Simplified this so the panic/error is not necessary.

File: s3proxy/cmd/main.go
Comment (daniel-weisse) [inline s3proxy/cmd/main.go:57]:
```suggestion
	log.With(zap.String("ip", flags.ip), zap.Bool("tlsEnabled", !flags.noTLS), zap.String("region", flags.Region)).Infof("listening")
```

**Review by daniel-weisse — APPROVED** [review]

**Review by 3u13r — CHANGES_REQUESTED** [review]

File: s3proxy/cmd/main.go
Comment (derpsteb) [reply s3proxy/cmd/main.go]: s3proxy now uses port 4433. Non-configurable for now. The deployment's service remaps that to 443.

**Review by 3u13r — APPROVED** [review]
LGTM, should we add e2e test tickets to the backlog?

**Comment by derpsteb** [comment]
Yes. Put into AB#3479 and AB#3480.

**Pull request merged**
